src/hotspot/share/runtime/continuationFreezeThaw.cpp

```diff
@@ -1965,6 +1965,9 @@ inline void ThawBase::patch(frame& f, const frame& caller, bool bottom) {
   if (bottom) {
     ContinuationHelper::Frame::patch_pc(caller, _cont.is_empty() ? caller.pc()
                                                                  : StubRoutines::cont_returnBarrier());
+  } else {
+    // caller might have been deoptimized during thaw but we've overwritten the return address when copying f
+    ContinuationHelper::Frame::patch_pc(caller, caller.raw_pc());
```

Comment:
What is raw_pc() vs pc() ?  why? is _cont.is_empty() true here?

Reply:
raw_pc is the "physical" return pc, and could point to the deopt handler; pc is the original caller address whether or not the caller is deoptimized. So raw_pc and pc are different iff the frame is deoptimized.
The relevant background is in the (pre-Loom) deoptimization-related frame methods, raw_pc and patch_pc (e.g. they say that deoptimizing stores the original pc in the frame).

Reply:
Looking at JDK 18 runtime code for raw_pc() (should probably change the comment since it's not only used by deoptimization anymore): I'm told pc_return_offset is zero so the 'else' clause doesn't change the caller.pc if the frame isn't deoptimized.
Thanks to @pchilano also for explanation.
Can you change the comment to:
s/when copying f/\nwhen copying the caller frame from the heap. The caller.pc is unchanged otherwise./

Reply:
it's not 'f'.

Reply:
Changed to say "when copying f from the heap. The caller.pc is unchanged otherwise" because it's copying f, the callee, that's the issue, as the return address to the caller is in the callee frame.

Reply:
Ok. didn't see 'f' in the diffs.